solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java

```diff
@@ -113,6 +118,25 @@ protected boolean wasCommError(Throwable rootCause) {
     return false;
   }
 
+  /**
+   * Execute an asynchronous request against a Solr server for a given collection
+   *
+   * @param request the request to execute
+   * @param collection the collection to execute the request against
+   * @param asyncListener the request listener, not null (use a no-op listener instead for async requests with no callback)
+   *
+   * @return a {@link Cancellable} allowing you to cancel execution of the async request
+   *
+   * @throws IOException If there is a low-level I/O error.
+   * @throws SolrServerException if there is an error on the server
+   */
+  public Cancellable asyncRequest(@SuppressWarnings({"rawtypes"}) SolrRequest request,
```

Comment:
Maybe instead of returning Cancellable and doing the hacks to make that work, and instead of the caller passing AsyncListener, perhaps the callback argument should be a `CompletableFuture<NamedList<Object>>` ?  This has a cancel method, and allows a single type to handle multiple use-cases that is also likely to be known to callers than yet more Solr-isms.

Reply:
I'm curious what you think @CaoManhDat

Reply:
Hmm, `CompletableFuture<NamedList<Object>>` along with with listener, it seems tricky to me.

Reply:
To be clear, I am not proposing both (CF & listener), I'm proposing just CF (of NamedList generic type).  The CF is versatile enough to allow the client to (1) cancel if desired, (2) to provide a Future; maybe the client prefers a Future (3) to provide a listener for just-in-time action *if* he/she chooses (e.g. via cf.handle(....)).

tricky == fun :-). I just spent some time playing around with a CF and the tricky part seems propagating a just-in-time cancel from the client calling cf.cancel() onwards to something else that can do something immediately, thus avoiding a poll.

Reply:
Some play code:
```
  public static void main(String[] args) throws InterruptedException, ExecutionException {
    CompletableFuture<String> clientCf = new CompletableFuture<>();
    final Thread thread = new Thread(() -> {
      CompletableFuture<String> workCf = clientCf.whenComplete((result, throwable) -> {
        System.out.println("whenComplete r:" + result + " t:" + throwable);
      });

      try {
        Thread.sleep(2000);
        System.out.println("bg: cancelled? " + workCf.isCancelled());

        clientCf.complete("Completed-Result");
      } catch (InterruptedException e) {
        System.out.println("bg: Interrupted");
        e.printStackTrace();
      }
    });
    thread.start();

    Thread.sleep(1000);
    boolean cancelRsp = clientCf.cancel(true);
    System.out.println("client: cancelRsp (was cancelled)="+cancelRsp);
    if (!cancelRsp) {
      System.out.println("client: " + clientCf.get());
    }

    thread.join();
  }
```